File: frontend/src/utils/refreshInterval.ts
import { showAsFps } from '../decorators/refreshInterval'
import type { FrameScene, StateField } from '../types'

/**
 * The scene's refresh interval as a state field.
 *
 * Every scene has one state key that holds "seconds between renders":
 *
 * - a field marked `role: 'refreshInterval'` (the first one wins), else
 * - a float or integer field literally named `refreshInterval`, else
 * - an implicit public field `refreshInterval`, seeded from
 *   `settings.refreshInterval` and appended after the scene's own fields, so
 *   every surface that shows a scene's options ends with "Refresh interval
 *   (seconds)".
 *
 * A field the scene declares stays where the scene put it: declaring it is how
 * an author decides where the control goes, what it is called, and — with
 * `access: 'private'` — that people do not get one. A `refreshInterval` field
 * of any other type is the scene using the name for something else: it is left
 * alone and no implicit field is added over it (`key` is then empty).
 *
 * The runtime applies the same rules (`frameos/src/frameos/refresh_interval.nim`,
 * and `backend/app/utils/refresh_interval.py` for compiled scenes);
 * `frameos/wasm/src/refreshInterval.ts` is the preview package's copy. Keep
 * them in step.
 */

export const REFRESH_INTERVAL_ROLE = 'refreshInterval'
export const REFRESH_INTERVAL_FIELD_NAME = 'refreshInterval'
export const REFRESH_INTERVAL_LABEL = 'Refresh interval (seconds)'
export const DEFAULT_REFRESH_INTERVAL_SECONDS = 300

type SceneLike = Pick<FrameScene, 'fields' | 'settings'> | null | undefined

/** A positive, finite number of seconds, or 0 for anything else. */
export function parseRefreshSeconds(value: unknown): number {
  if (typeof value !== 'number' && typeof value !== 'string') {
    return 0
  }
  if (typeof value === 'string' && value.trim() === '') {
    return 0
  }
  const seconds = Number(value)
  return Number.isFinite(seconds) && seconds > 0 ? seconds : 0
}

function isNumericField(field: StateField): boolean {
  return field.type === 'float' || field.type === 'integer'
}

/**
 * Index of the field that carries the refresh interval; -1 when the scene
 * declares none. A role beats the name, and the name only counts on a numeric field.
 */
export function refreshIntervalFieldIndex(fields: StateField[] | null | undefined): number {
  const list = fields ?? []
  const byRole = list.findIndex((field) => field?.role === REFRESH_INTERVAL_ROLE)
  if (byRole >= 0) {
    return byRole
  }
  return list.findIndex((field) => field?.name === REFRESH_INTERVAL_FIELD_NAME && isNumericField(field))
}

function settingsDefault(scene: SceneLike): number {
  return parseRefreshSeconds(scene?.settings?.refreshInterval) || DEFAULT_REFRESH_INTERVAL_SECONDS
}

export function implicitRefreshIntervalField(defaultSeconds: number): StateField {
  return {
    name: REFRESH_INTERVAL_FIELD_NAME,
    label: REFRESH_INTERVAL_LABEL,
    type: 'float',
    value: String(defaultSeconds),
    persist: 'disk',
    access: 'public',
    role: REFRESH_INTERVAL_ROLE,
  }
}

export interface ResolvedRefreshInterval {
  /** The scene's fields in their own order, plus the implicit one last. */
  fields: StateField[]
  /** The state key that holds the interval ('' = settings only). */
  key: string
  /** What the scene ships with. */
  defaultSeconds: number
  /** The field was added here, not declared by the scene. */
  implicit: boolean
}

export function resolveRefreshInterval(scene: SceneLike): ResolvedRefreshInterval {
  const fields = (scene?.fields ?? []).filter((field): field is StateField => !!field && typeof field === 'object')
  const index = refreshIntervalFieldIndex(fields)
  const declared = index >= 0 ? fields[index] : undefined
  if (declared) {
    return {
      // Stamp the role on a field taken over by name, so a surface only ever
      // has to ask `field.role === 'refreshInterval'`.
      fields: fields.map((field, i) => (i === index ? { ...field, role: REFRESH_INTERVAL_ROLE } : field)),
      key: declared.name,
      defaultSeconds: parseRefreshSeconds(declared.value) || settingsDefault(scene),
      implicit: false,
    }
  }
  const defaultSeconds = settingsDefault(scene)
  if (fields.some((field) => field.name === REFRESH_INTERVAL_FIELD_NAME)) {
    // The name is taken by a field of another type: settings only.
    return { fields, key: '', defaultSeconds, implicit: false }
  }
  return {
    fields: [...fields, implicitRefreshIntervalField(defaultSeconds)],
    key: REFRESH_INTERVAL_FIELD_NAME,
    defaultSeconds,
    implicit: true,
  }
}

/** Every state field of the scene in its own order, plus the implicit refresh interval last when it has none. */
export function sceneStateFields(scene: SceneLike): StateField[] {
  return resolveRefreshInterval(scene).fields
}

/** The fields a control surface offers: the public ones of sceneStateFields(). */
export function scenePublicStateFields(scene: SceneLike): StateField[] {
  return sceneStateFields(scene).filter((field) => field.access !== 'private')
}

/** "1 hour", "15 min", "24 fps": what a number of seconds means, for the hint next to the input. */
export function describeRefreshSeconds(value: unknown): string {
  const seconds = parseRefreshSeconds(value)
  return seconds > 0 ? showAsFps(seconds) : ''
}

/** The scene's seconds between renders, given the state it currently holds (if known). */
export function sceneRefreshSeconds(scene: SceneLike, state?: Record<string, unknown> | null): number {
  const { key, defaultSeconds } = resolveRefreshInterval(scene)
  return (key ? parseRefreshSeconds(state?.[key]) : 0) || defaultSeconds
}

File: backend/app/utils/refresh_interval.py
"""The scene's refresh interval as a state field.

Every scene has one state key that holds "seconds between renders": a field
with ``role: "refreshInterval"`` (the first one wins), else a float or integer
field literally named ``refreshInterval``, else an implicit public field seeded
from ``settings.refreshInterval`` and appended after the scene's own fields.

A field the scene declares stays where the scene put it. A ``refreshInterval``
field of any other type is the scene using the name for something else: it is
left alone and no implicit field is added over it (``key`` is then empty).

The same rules live in ``frameos/src/frameos/refresh_interval.nim`` (the
runtime) and ``frontend/src/utils/refreshInterval.ts`` (every control
surface); keep them in step.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any

REFRESH_INTERVAL_ROLE = "refreshInterval"
REFRESH_INTERVAL_FIELD_NAME = "refreshInterval"
REFRESH_INTERVAL_LABEL = "Refresh interval (seconds)"
DEFAULT_REFRESH_INTERVAL_SECONDS = 300.0


def parse_refresh_seconds(value: Any) -> float:
    """A positive, finite number of seconds, or 0 for anything else."""
    if isinstance(value, bool) or value is None:
        return 0.0
    try:
        seconds = float(value.strip() if isinstance(value, str) else value)
    except (TypeError, ValueError):
        return 0.0
    return seconds if math.isfinite(seconds) and seconds > 0 else 0.0


NUMERIC_FIELD_TYPES = ("float", "integer")


def refresh_interval_field_index(fields: list[dict]) -> int:
    for index, field in enumerate(fields):
        if isinstance(field, dict) and field.get("role") == REFRESH_INTERVAL_ROLE:
            return index
    for index, field in enumerate(fields):
        if (
            isinstance(field, dict)
            and field.get("name") == REFRESH_INTERVAL_FIELD_NAME
            and field.get("type") in NUMERIC_FIELD_TYPES
        ):
            return index
    return -1


@dataclass
class ResolvedRefreshInterval:
    fields: list[dict]  # the scene's fields in their own order, plus the implicit one last
    key: str  # the state key that holds the interval ("" = settings only)
    default_seconds: float  # what the scene ships with
    implicit: bool  # the field was added here, not declared by the scene


def resolve_refresh_interval(fields: list[dict] | None, settings_seconds: Any) -> ResolvedRefreshInterval:
    fields = [field for field in (fields or []) if isinstance(field, dict)]
    settings_default = parse_refresh_seconds(settings_seconds) or DEFAULT_REFRESH_INTERVAL_SECONDS
    index = refresh_interval_field_index(fields)
    if index >= 0:
        field = fields[index]
        declared = parse_refresh_seconds(field.get("value"))
        return ResolvedRefreshInterval(fields, str(field.get("name", "")), declared or settings_default, False)
    if any(field.get("name") == REFRESH_INTERVAL_FIELD_NAME for field in fields):
        # The name is taken by a field of another type: settings only.
        return ResolvedRefreshInterval(fields, "", settings_default, False)
    implicit = {
        "name": REFRESH_INTERVAL_FIELD_NAME,
        "label": REFRESH_INTERVAL_LABEL,
        "type": "float",
        # "900", not "900.0": the same text the frontend seeds the input with
        "value": str(int(settings_default)) if settings_default.is_integer() else str(settings_default),
        "persist": "disk",
        "access": "public",
        "role": REFRESH_INTERVAL_ROLE,
    }
    return ResolvedRefreshInterval([*fields, implicit], REFRESH_INTERVAL_FIELD_NAME, settings_default, True)
